Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp

```diff
@@ -90,9 +90,59 @@ Bytecode::CodeGenerationErrorOr<Optional<ScopedOperand>> BinaryExpression::gener
         return dst;
     }
 
-    auto lhs = TRY(m_lhs->generate_bytecode(generator)).value();
-    auto rhs = TRY(m_rhs->generate_bytecode(generator)).value();
+    // OPTIMIZATION: If LHS and/or RHS are numeric literals, we make sure they are converted to i32/u32
+    //               as appropriate, to avoid having to perform these conversions at runtime.
+
+    auto get_left_side = [&](Expression const& side) -> CodeGenerationErrorOr<Optional<ScopedOperand>> {
+        switch (m_op) {
+        case BinaryOp::BitwiseAnd:
+        case BinaryOp::BitwiseOr:
+        case BinaryOp::BitwiseXor:
+        case BinaryOp::LeftShift:
+        case BinaryOp::RightShift:
+        case BinaryOp::UnsignedRightShift:
+            // LHS will always be converted to i32 for these ops.
+            if (side.is_numeric_literal()) {
```

Comment:
doesn't the same/similar hold true for numeric string literals?
(Or any contextually convertible literal, like some arrays/objects as well)